### PR TITLE
kvs: support append operation

### PR DESCRIPTION
**Pull request by chu11**
This PR implements basic append support in the KVS and a basic ```--append``` option in flux-kvs.

### CODE REVIEW

File: src/common/libkvs/kvs_txn.c [hunk 1]
```diff
@@ -122,7 +122,7 @@ int flux_kvs_txn_put_raw (flux_kvs_txn_t *txn, int flags,
         errno = EINVAL;
         goto error;
     }
-    if (validate_flags (flags, 0) < 0)
+    if (validate_flags (flags, FLUX_KVS_APPEND) < 0)
```
Comment (garlick) [inline src/common/libkvs/kvs_txn.c]: We should allow append in `flux_kvs_txn_put()` also.

File: src/common/libkvs/kvs_txn.h [hunk 2]
```diff
@@ -7,6 +7,10 @@ extern "C" {
 
 #include <stdarg.h>
 
+enum {
```
Comment (garlick) [inline src/common/libkvs/kvs_txn.h]: Just noticed all the flags in kvs.h are now lookup specific.  I moved them into kvs.h because earlier we alloed the FLUX_KVS_TREEOBJ flag with txn_put.

I guess I'd vote we put all the KVS user flags in the same enum in kvs.h unless others feel they should be split up.

File: src/modules/kvs/commit.c [hunk 3]
```diff
@@ -369,6 +376,116 @@ static int commit_unroll (commit_t *c, int current_epoch, json_t *dir)
...
+        errno = EOPNOTSUPP;
+        return -1;
+    }
+    else if (treeobj_is_dir (entry)
+             || treeobj_is_dirref (entry)) {
+        errno = EISDIR;
+        return -1;
+    }
+    else {
+        char *s = json_dumps (entry, 0);
```
Comment (garlick) [inline src/modules/kvs/commit.c]: use treeobj_encode()?

File: doc/man1/flux-kvs.adoc [hunk 4]
```diff
@@ -50,13 +50,15 @@ interpreted as encoded JSON and formatted accordingly.  If '-r', value
...
-*put* [-j|-r|-t] [-n] 'key=value' ['key=value...']::
+*put* [-j|-r|-t] [-n] [-A] 'key=value' ['key=value...']::
 Store 'value' under 'key' and commit it.  If it already has a value,
 overwrite it.  If no options, value is stored directly.  If '-j', it is
 first encoded as JSON, then stored.  If '-r', the value may be read from
 standard input if specified as "-", and may include embedded NULL bytes.
 If '-t', value is stored as a RFC 11 object.  '-n' prevents the commit
-from being merged with with other contemporaneous commits.
+from being merged with with other contemporaneous commits.  '-A' appends the
+value to a key instead of overwriting the value.  Append only works with the
```
Comment (garlick) [inline doc/man1/flux-kvs.adoc]: should work with no options, if my suggestion is accepted.   JSON is the one that gets weird since a JSON object appended to a JSON object isn't valid JSON.

File: src/cmd/flux-kvs.c [hunk 5]
```diff
@@ -94,6 +94,9 @@ static struct optparse_option put_opts[] =  {
     { .name = "no-merge", .key = 'n', .has_arg = 0,
       .usage = "Set the NO_MERGE flag to ensure commit is standalone",
     },
+    { .name = "append", .key = 'A', .has_arg = 0,
```
Comment (garlick) [inline src/cmd/flux-kvs.c]: Any reason not to use little a for the short opt?

Comment (chu11) [reply src/cmd/flux-kvs.c]: little a is used for the "at" option throughout flux kvs, so picked big A just to be different.

File: doc/man1/flux-kvs.adoc (hunk 4 above)
Comment (chu11) [reply doc/man1/flux-kvs.adoc]: Ahh, I didn't adjust given changes with #1262.  Will have to go back and change that.

File: src/common/libkvs/kvs_txn.h (hunk 2 above)
Comment (chu11) [reply src/common/libkvs/kvs_txn.h]: ahh, sounds good to put it all in kvs.h

File: src/modules/kvs/commit.c (hunk 3 above)
Comment (chu11) [reply src/modules/kvs/commit.c]: Since this is the "corrupt treeobj" path, I think using json_dumps() is the right idea.  

Using treeobj_encode() works right now b/c it presently just calls json_dumps() directly and does nothing else.  Could treeobj_encode() ever check for treeobj_validate?

**Comment by chu11** [comment]
repushed with minor fixes as discussed above.  Fixed a major bug I found.  Required refactoring of ```treeobj_copy_dir()``` into ```treeobj_copy()``` and working w/ valrefs and dirrefs too.

File: src/cmd/flux-kvs.c (hunk 5 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Oh duh.

File: doc/man1/flux-kvs.adoc [hunk 6]
```diff
@@ -57,8 +57,8 @@ first encoded as JSON, then stored.  If '-r', the value may be read from
 standard input if specified as "-", and may include embedded NULL bytes.
 If '-t', value is stored as a RFC 11 object.  '-n' prevents the commit
 from being merged with with other contemporaneous commits.  '-A' appends the
-value to a key instead of overwriting the value.  Append only works with the
-'-r' option.
+value to a key instead of overwriting the value.  Append works with no options
```
Comment (garlick) [inline doc/man1/flux-kvs.adoc]: Maybe simpler to say

> Append is incompatible with the `-j` option.

File: src/common/libkvs/treeobj.h [hunk 7]
```diff
@@ -69,12 +69,14 @@ json_t *treeobj_get_entry (json_t *obj, const char *name);
...
- * on dir objects, as the latter is not safe on treeobj objects.
+/* Copy a treeobj.
+ * This function performs a shallow copy on a dir, dirref, or valref
+ * object.  Use this function over json_copy() on dir, dirref, or
+ * valref objects, as the latter is not safe on treeobj objects.
+ * This function will copy one additional level deeper in the treeobj
+ * so it is safe to manipulate.
  */
-json_t *treeobj_copy_dir (json_t *obj);
+json_t *treeobj_copy (json_t *obj);
```
Comment (garlick) [inline src/common/libkvs/treeobj.h]: Should this be named so that it's more clear that it makes a thin copy of the `data` object?

This property is critical to understanding how commits are processed.

Maybe `treeobj_copy_thin()` or similar?

**Comment by garlick** [comment]
Good job finding that bug!

Maybe the json objects returned from the cache should be tagged `const` ?

File: src/common/libkvs/treeobj.h (hunk 7 above)
Comment (garlick) [reply src/common/libkvs/treeobj.h]: Actually it is a deep copy as long as it's not a directory that includes "rolled up" subdirectories.  Hmm, maybe just need to add a new `treeobj_copy()` that makes a deep copy and leave that other one alone?

**Comment by chu11** [comment]
good idea on making the cache returns ```const```.  I'll instead add that to the list of cleanups in #1264.

File: src/common/libkvs/treeobj.h (hunk 7 above)
Comment (chu11) [reply src/common/libkvs/treeobj.h]: I can see your point that ```treeobj_copy()``` by itself may be ambiguous.  I had wanted to avoid it, perhaps adding separate ```treeobj_copy_dirref()``` and ```treeobj_copy_valref()``` functions would be better?

**Comment by garlick** [comment]
Don't forget to update the FLAGS section of flux_kvs_txn_create.adoc with a description of FLUX_KVS_APPEND.

File: src/common/libkvs/treeobj.h (hunk 7 above)
Comment (garlick) [reply src/common/libkvs/treeobj.h]: How about `treeobj_copy()` that fully copies any treeobj, and then rename `treeobj_copy_dir()` to something more indicitave of what it's doing like `treeobj_copy_dir_thin()`?  (the rename could be deferred to another PR)

**Comment by chu11** [comment]
just pushed with smaller incremental changes as discussed above.  Also completely redid the prior ```treeobj_copy_dir()``` refactor.  Instead, a new function ```treeobj_copy()``` does deep copies on all tree objs.

File: src/common/libkvs/treeobj.h (hunk 7 above)
Comment (chu11) [reply src/common/libkvs/treeobj.h]: Now that I think about it, how about just making a function ```treeobj_copy_shallow()```.  It can be made quite clear that it's a shallow copy of the data within a treeobj, regardless of the data it is.

File: src/common/libkvs/treeobj.c [hunk 8]
```diff
@@ -263,6 +263,84 @@ int treeobj_insert_entry (json_t *obj, const char *name, json_t *obj2)
     return 0;
 }
 
+json_t *treeobj_copy (json_t *obj)
```
Comment (garlick) [inline src/common/libkvs/treeobj.c]: would `json_deep_copy()` work here?

Comment (chu11) [reply src/common/libkvs/treeobj.c]: naming wise?  That would follow jansson style naming.  Are you thinking ```treeobj_copy()``` would be the shallow equivalent?

Comment (garlick) [reply src/common/libkvs/treeobj.c]: I meant could the jansson function `json_deep_copy()` do what `treeobj_copy()` proposes to do here?  

http://jansson.readthedocs.io/en/2.8/apiref.html#copying

Comment (chu11) [reply src/common/libkvs/treeobj.c]: ohhh, duh!  You're right.  I'm going to rename to ```treeobj_deep_copy()``` as well, as that would follow the jansson style.  ```treeobj_copy()``` will be the treeobj equivalent of shallow copy.

**Comment by chu11** [comment]
just re-pushed, decided to rename functions, so now we have ```treeobj_copy()``` which is a treeobj shallow copy, and ```treeobj_deep_copy()```, which is a deep copy.

Also noticed a corner case in ```treeobj_create_symlink()```, so I added a check and unit test there.

**Comment by chu11** [comment]
I occurred to me on the drive home.  Given how I implemented ```treeobj_copy()``` as a shallow copy of the data within a treeobject, ```treeobj_deep_copy()``` is no longer needed.  Should we keep it around as a potentially useful future function?  Or axe it.

**Comment by garlick** [comment]
> Given how I implemented treeobj_copy() as a shallow copy of the data within a treeobject, treeobj_deep_copy() is no longer needed. Should we keep it around as a potentially useful future function? Or axe it.

I may be getting lost here.  Why not just have a `treeobj_copy()` that directly calls `json_deep_copy()`, and the original `treeobj_copy_dir()` (possibly renamed) with its specialized function for commit processing?  I'm not sure why `treeobj_copy()` needs to be type specific.

**Comment by chu11** [comment]
I don't think we're on the same page on this.  I'll drop by when I get in.

**Comment by chu11** [comment]
just re-pushed, cleanup ```treeobj_copy()``` by calling ```json_deep_copy()```.  Also nit cleanup in ```treeobj_deep_copy()```.

**Comment by chu11** [comment]
Ugh ...I forgot to commit --amend something.  tiny re-push.

**Comment by garlick** [comment]
OK, ready to merge?

**Comment by chu11** [comment]
just squashed and re-pushed.  Two tiny tweaks in the squash.

Removed calls to ```treeobj_validate()``` in ```treeobj_copy()``` and ```treeobj_deepcopy()``` as they were considered superfluous.

Also removed a few unit tests for ```treeobj_deep_copy()```, as they were a bit superfluous given the tests that existed for ```treeobj_copy()```.

**Comment by garlick** [comment]
Looks good @chu11 - will merge once travis passes.

**Comment by chu11** [comment]
hit

```ERROR: t2001-jsc.t - exited with status 137 (terminated by signal 9?)```

i'm assuming slowness and a broker had to be killed.  restarting build.
